### PR TITLE
Migrate the ability to upload and download binlog to cpp

### PR DESCRIPTION
**Pull request by xige-16**
fix issue: #21266
/kind bug
Signed-off-by: xige-16 <xi.ge@zilliz.com>

### CODE REVIEW

File: internal/util/dependency/factory.go [hunk 1]
```diff
@@ -24,6 +24,15 @@ func NewDefaultFactory(standAlone bool) *DefaultFactory {
 	}
 }
 
+// Only for test
+func NewDefaultFactoryV2(standAlone bool, params *paramtable.ComponentParam) *DefaultFactory {
```
Comment (czs007) [inline internal/util/dependency/factory.go:28]: It is recommended to include the words Test or Mock in the name

File: internal/core/src/storage/CMakeLists.txt [hunk 2]
```diff
@@ -36,25 +36,17 @@ set(STORAGE_FILES
...
 add_library(milvus_storage SHARED ${STORAGE_FILES})
 
 find_package(Boost REQUIRED COMPONENTS filesystem)
 
-if(BUILD_DISK_ANN STREQUAL "ON")
-    target_link_libraries(milvus_storage PUBLIC milvus_common Boost::filesystem aws-cpp-sdk-s3 pthread)
-else()
-    target_link_libraries(milvus_storage PUBLIC milvus_common Boost::filesystem pthread)
-endif()
+target_link_libraries(milvus_storage PUBLIC milvus_common Boost::filesystem aws-cpp-sdk-s3 pthread)
```
Comment (czs007) [inline internal/core/src/storage/CMakeLists.txt:49]: always link aws-cp-sdk-s3 ?

File: internal/core/src/storage/DiskFileManagerImpl.cpp [hunk 3]
```diff
@@ -118,31 +70,75 @@ DiskFileManagerImpl::AddFile(const std::string& file) noexcept {
...
 
+void
+DiskFileManagerImpl::AddBatchIndexFiles(const std::string& local_file_name,
+                                        const std::vector<int64_t>& local_file_offsets,
+                                        const std::vector<std::string>& remote_files,
+                                        const std::vector<int64_t>& remote_file_sizes) {
+    auto& local_chunk_manager = LocalChunkManager::GetInstance();
+    auto& pool = ThreadPool::GetInstance();
+
+    auto LoadIndexFromDisk = [&](const std::string& file, const int64_t offset,
```
Comment (xiaofan-luan) [inline internal/core/src/storage/DiskFileManagerImpl.cpp:113]: might not be a good idea to load all data into memory in one run

File: internal/util/indexcgowrapper/index.go [hunk 4]
```diff
@@ -339,3 +326,38 @@ func (index *CgoIndex) CleanLocalData() error {
 	status := C.CleanLocalData(index.indexPtr)
```
Comment (xiaofan-luan) [inline internal/util/indexcgowrapper/index.go:326]: 1. seems cleanLocal Data is not actually used in index building path.
2. what happened if clean failed? the garbage data is leaved there, would that affect next index build or load?
3.

**Comment by czs007** [comment]
will be shipped in 2.2.5

File: internal/indexnode/indexnode_service_test.go [hunk 5]
```diff
@@ -30,133 +30,133 @@ func genStorageConfig() *indexpb.StorageConfig {
...
-	t.Run("drop not exists jobs", func(t *testing.T) {
-		status, err := in.DropJobs(ctx, &indexpb.DropJobsRequest{
-			ClusterID: clusterID,
-			BuildIDs:  []int64{100001},
-		})
-		assert.Nil(t, err)
-		assert.Equal(t, status.ErrorCode, commonpb.ErrorCode_Success)
-	})
-}
+//func TestIndexNodeSimple(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (xige-16) [reply internal/indexnode/indexnode_service_test.go:33]: cpp ut can't link to minio in the github ut environment before, but it has been solved now, in another pr https://github.com/milvus-io/milvus/pull/22783

File: internal/indexnode/task.go [hunk 6]
```diff
@@ -242,208 +239,121 @@ func (it *indexBuildTask) LoadData(ctx context.Context) error {
...
+		if err != nil {
+			log.Ctx(ctx).Warn("IndexNode get local used size failed")
+			return err
+		}
+		fieldDataSize, err := estimateFieldDataSize(it.statistic.Dim, it.req.GetNumRows(), it.fieldType)
+		if err != nil {
+			log.Ctx(ctx).Warn("IndexNode get local used size failed")
+			return err
+		}
+		usedLocalSizeWhenBuild := int64(float64(fieldDataSize)*diskUsageRatio) + localUsedSize
```
[the remark this answers is not in the document]
Comment (xige-16) [reply internal/indexnode/task.go:269]: From knowhere's test result

File: internal/querynode/segment.go [hunk 7]
```diff
@@ -781,35 +781,113 @@ func (s *Segment) segmentDelete(offset int64, entityIDs []primaryKey, timestamps
...
+
+func (s *Segment) LoadMultiFieldData(rowCount int64, fields []*datapb.FieldBinlog) error {
+	s.mut.RLock()
+	defer s.mut.RUnlock()
+	if !s.healthy() {
+		return fmt.Errorf("%w(segmentID=%d)", ErrSegmentUnhealthy, s.segmentID)
+	}
+
+	loadFieldDataInfo, err := newLoadFieldDataInfo()
+	defer deleteFieldDataInfo(loadFieldDataInfo)
```
[the remark this answers is not in the document]
Comment (xige-16) [reply internal/querynode/segment.go:833]: It is perfectly safe to delete a null pointer in C++

**Comment by xiaofan-luan** [comment]
you have to change the commit message. 
currently is shows merged from XXX

File: internal/querynode/segment.go [hunk 8]
```diff
@@ -781,35 +781,113 @@ func (s *Segment) segmentDelete(offset int64, entityIDs []primaryKey, timestamps
...
+func (s *Segment) LoadFieldData(fieldID int64, rowCount int64, field *datapb.FieldBinlog) error {
 	s.mut.RLock()
 	defer s.mut.RUnlock()
 	if !s.healthy() {
 		return fmt.Errorf("%w(segmentID=%d)", ErrSegmentUnhealthy, s.segmentID)
 	}
 
-	dataBlob, err := proto.Marshal(data)
+	loadFieldDataInfo, err := newLoadFieldDataInfo()
+	defer deleteFieldDataInfo(loadFieldDataInfo)
```
[the remark this answers is not in the document]
Comment (xige-16) [reply internal/querynode/segment.go:873]: It is perfectly safe to delete a null pointer in C++

File: internal/core/src/index/ScalarIndexSort-inl.h [hunk 9]
```diff
@@ -24,16 +24,36 @@
...
 #include "common/Utils.h"
 #include "common/Slice.h"
+#include "index/Utils.h"
 
 namespace milvus::index {
 
 template <typename T>
-inline ScalarIndexSort<T>::ScalarIndexSort() : is_built_(false), data_() {
+inline ScalarIndexSort<T>::ScalarIndexSort(storage::FileManagerImplPtr file_manager) : is_built_(false), data_() {
+    if (file_manager != nullptr) {
```
[the remark this answers is not in the document]
Comment (xige-16) [reply internal/core/src/index/ScalarIndexSort-inl.h:33]: fixed

File: internal/core/src/index/ScalarIndexSort-inl.h [hunk 10]
```diff
@@ -98,6 +134,16 @@ ScalarIndexSort<T>::Load(const BinarySet& index_binary, const Config& config) {
     is_built_ = true;
 }
 
+template <typename T>
+inline void
+ScalarIndexSort<T>::Load(const Config& config) {
+    auto index_files = GetValueFromConfig<std::vector<std::string>>(config, "index_files");
+    AssertInfo(index_files.has_value(), "index file paths is empty when load disk ann index");
+    auto binary_set = file_manager_->CacheIndexToMemory(index_files.value());
```
[the remark this answers is not in the document]
Comment (xige-16) [reply internal/core/src/index/ScalarIndexSort-inl.h]: fixed

File: internal/core/src/storage/DiskFileManagerImpl.cpp [hunk 11]
```diff
@@ -233,16 +209,51 @@ DiskFileManagerImpl::CacheBatchIndexFilesToDisk(const std::vector<std::string>&
 }
 
 std::string
-DiskFileManagerImpl::GetFileName(const std::string& localfile) {
-    boost::filesystem::path localPath(localfile);
-    return localPath.filename().string();
+DiskFileManagerImpl::CacheRawDataToDisk(std::vector<std::string> remote_files) {
```
[the remark this answers is not in the document]
Comment (xige-16) [reply internal/core/src/storage/DiskFileManagerImpl.cpp:217]: yes, the cleanLocalData func would be called in golang

File: internal/core/src/storage/DiskFileManagerImpl.cpp [hunk 12]
```diff
@@ -118,31 +70,75 @@ DiskFileManagerImpl::AddFile(const std::string& file) noexcept {
...
+            batch_remote_files.clear();
+            remote_file_sizes.clear();
+            local_file_offsets.clear();
+        }
+
         auto batch_size = std::min(index_file_slice_size << 20, int64_t(fileSize) - offset);
-        // Put file to remote
-        char objectKey[200];
-        snprintf(objectKey, sizeof(objectKey), "%s/%s_%d", remotePrefix.c_str(), fileName.c_str(), slice_num);
+        batch_remote_files.emplace_back(remotePrefix + "/" + fileName + "_" + std::to_string(slice_num));
```
[the remark this answers is not in the document]
Comment (xige-16) [reply internal/core/src/storage/DiskFileManagerImpl.cpp]: fixed

File: internal/util/indexcgowrapper/index.go (hunk 4 above)
Comment (xige-16) [reply internal/util/indexcgowrapper/index.go:326]: 1. fixed
2. Even the garbage data is leaved there, don't  affect next index build or load. The local index and raw data files will be removed when rebuild or load
 https://github.com/milvus-io/milvus/blob/fce815e4145b279eda6fdf1aa3367153cf0a9e6a/internal/core/src/index/VectorDiskIndex.cpp#L34-L48

File: internal/util/indexcgowrapper/build_index_info.go [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+}
+
+func (bi *BuildIndexInfo) AppendBuildTypeParam(typeParams map[string]string) error {
+	protoTypeParams := &indexcgopb.TypeParams{
+		Params: make([]*commonpb.KeyValuePair, 0),
+	}
+	for key, value := range typeParams {
+		protoTypeParams.Params = append(protoTypeParams.Params, &commonpb.KeyValuePair{Key: key, Value: value})
+	}
+	typeParamsStr := proto.MarshalTextString(protoTypeParams)
```
[the remark this answers is not in the document]
Comment (xige-16) [reply internal/util/indexcgowrapper/build_index_info.go]: fixed by using proto.Marshal

File: internal/core/src/segcore/SegmentSealedImpl.cpp [hunk 14]
```diff
@@ -142,13 +144,29 @@ SegmentSealedImpl::LoadScalarIndex(const LoadIndexInfo& info) {
...
-SegmentSealedImpl::LoadFieldData(const LoadFieldDataInfo& info) {
+SegmentSealedImpl::LoadFieldData(const LoadFieldDataInfo& load_info) {
     // print(info);
     // NOTE: lock only when data is ready to avoid starvation
-    AssertInfo(info.row_count > 0, "The row count of field data is 0");
-    auto field_id = FieldId(info.field_id);
-    AssertInfo(info.field_data != nullptr, "Field info blob is null");
-    auto size = info.row_count;
+    for (auto& field_info : load_info.field_infos) {  // only one field for now, parallel load field data in golang
+        auto info = field_info.second;
```
Comment (yah01) [inline internal/core/src/segcore/SegmentSealedImpl.cpp:151]: use structured binding here to improve readability:
```
for (auto& [id, field_info] : load_info.field_infos)
```
also line 151 will copy the whole struct, which contains a vector of strings

**Review by yah01 — CHANGES_REQUESTED** [review]

**Review by yah01 — CHANGES_REQUESTED** [review]

File: internal/core/src/segcore/SegmentSealedImpl.cpp [hunk 15]
```diff
@@ -142,13 +144,29 @@ SegmentSealedImpl::LoadScalarIndex(const LoadIndexInfo& info) {
...
+        AssertInfo(info.row_count > 0, "The row count of field data is 0");
+        auto field_id = FieldId(info.field_id);
+        auto remote_chunk_manager = storage::RemoteChunkManagerFactory::GetInstance().GetRemoteChunkManager();
+        auto insert_files = info.insert_files;
```
[the remark this answers is not in the document]
Comment (xige-16) [reply internal/core/src/segcore/SegmentSealedImpl.cpp:155]: insert file need to be resorted. and the params info is const

File: internal/core/src/segcore/Utils.cpp [hunk 16]
```diff
@@ -34,6 +34,30 @@ ParsePksFromFieldData(std::vector<PkType>& pks, const DataArray& data) {
     }
 }
 
+void
+ParsePksFromFieldData(DataType data_type, std::vector<PkType>& pks, const std::vector<storage::FieldDataPtr>& datas) {
+    int64_t offset = 0;
+
+    for (auto& field_data : datas) {
```
Comment (xiaofan-luan) [inline internal/core/src/segcore/Utils.cpp:41]: try to avoid copies here

**Comment by xiaofan-luan** [comment]
I'm ok with the rest of the PR except the copy part

**Comment by xige-16** [comment]
rerun ut

**Comment by xige-16** [comment]
rerun ut

**Comment by xige-16** [comment]
rerun ut

**Pull request merged**